File: legacy-python/deadcode/visitor/ignore.py
import ast
from fnmatch import fnmatch, fnmatchcase
from functools import lru_cache
from pathlib import Path
from typing import Iterable, Set, Union

from deadcode.visitor.code_item import CodeItem


IGNORED_VARIABLE_NAMES = {'object', 'self'}
_PYTEST_FUNCTION_NAMES = {
    'setup_module',
    'teardown_module',
    'setup_function',
    'teardown_function',
}
_PYTEST_METHOD_NAMES = {
    'setup_class',
    'teardown_class',
    'setup_method',
    'teardown_method',
}
_PYTEST_FIXTURE_DECORATOR_NAMES = {
    '@pytest.fixture',
    '@pytest_asyncio.fixture',
    '@fixture',
}
PYTEST_USEFIXTURES_DECORATOR_NAMES = {
    '@pytest.mark.usefixtures',
    '@mark.usefixtures',
    '@usefixtures',
}
_OVERRIDE_DECORATOR_NAMES = {
    '@typing.override',
    '@typing_extensions.override',
    '@override',
}

ERROR_CODES = {
    'variable': b'DC01',
    'function': b'DC02',
    'class': b'DC03',
    'method': b'DC04',
    'attribute': b'DC05',
    'name': b'DC06',
    'import': b'DC07',
    'property': b'DC08',
    'unreachable_code': b'DC09',
    'empty_file': b'DC11',
    'commented_out_code': b'DC12',
    'ignore_expression': b'DC13',
}


def get_unused_items(defined_items: Iterable[CodeItem], used_names: Set[str]) -> Iterable[CodeItem]:
    unused_items = [item for item in defined_items if item.name not in used_names]
    unused_items.sort(key=lambda item: item.name.lower())
    return unused_items


def _is_special_name(name: str) -> bool:
    return name.startswith('__') and name.endswith('__')


def match(name: Union[str, Path], patterns: Iterable[str], case: bool = True) -> bool:
    func = fnmatchcase if case else fnmatch
    # .as_posix() (not str()) for Path names, so a pattern like '*/tests/*'
    # matches consistently regardless of the OS's native path separator.
    name_str = name.as_posix() if isinstance(name, Path) else name
    return any(func(name_str, pattern) for pattern in patterns)


def match_many(names: Union[Iterable[str], Iterable[Path]], patterns: Iterable[str], case: bool = True) -> bool:
    return any(match(name, patterns, case) for name in names)


@lru_cache(maxsize=None)
def _is_test_file(filename: Path) -> bool:
    # Called once per definition (function/class/method) in a file, so cache
    # per-filename: `filename.resolve()` is a filesystem syscall and the
    # answer never changes for repeated calls with the same file.
    return match(
        filename.resolve(),
        ['*/test/*', '*/tests/*', '*/test*.py', '*[-_]test.py'],
        case=False,
    )


def _is_conftest_file(filename: Path) -> bool:
    return filename.name == 'conftest.py'


def assigns_special_variable__all__(node: ast.Assign) -> bool:
    assert isinstance(node, ast.Assign)
    return isinstance(node.value, (ast.List, ast.Tuple)) and any(
        target.id == '__all__' for target in node.targets if isinstance(target, ast.Name)
    )


def ignore_class(filename: Path, class_name: str) -> bool:
    return _is_test_file(filename) and 'Test' in class_name


def ignore_import(filename: Path, import_name: str) -> bool:
    """
    Ignore star-imported names since we can't detect whether they are used.
    Ignore imports from __init__.py files since they're commonly used to
    collect objects from a package.
    """
    return filename.name == '__init__.py' or import_name == '*'


def ignore_function(filename: Path, function_name: str) -> bool:
    return (
        (function_name in _PYTEST_FUNCTION_NAMES or function_name.startswith('test_')) and _is_test_file(filename)
    ) or _ignore_pytest_hook(filename, function_name)


def _ignore_pytest_hook(filename: Path, function_name: str) -> bool:
    """
    `pytest_*` functions in conftest.py (e.g. `pytest_configure`,
    `pytest_collection_modifyitems`, `pytest_addoption`) are hook
    implementations that pytest calls automatically by name, never directly,
    so they would otherwise look unused.
    """
    return _is_conftest_file(filename) and function_name.startswith('pytest_')


def ignore_pytest_fixture(filename: Path, decorator_names: Iterable[str]) -> bool:
    """
    Pytest fixtures are consumed by name-matching (as a test's parameter, via
    `autouse=True`, `@pytest.mark.usefixtures`, or `request.getfixturevalue()`),
    so they're never called directly and would otherwise look unused.

    Trust any `@pytest.fixture`/`@pytest_asyncio.fixture` defined in conftest.py
    (auto-discovered project-wide by pytest) or in a recognized test file.
    Fixtures defined elsewhere must be exempted explicitly, e.g. via
    --ignore-names-if-decorated-with.
    """
    return (_is_conftest_file(filename) or _is_test_file(filename)) and match_many(
        decorator_names, _PYTEST_FIXTURE_DECORATOR_NAMES
    )


def ignore_override(decorator_names: Iterable[str]) -> bool:
    """
    `@typing.override` (and `@typing_extensions.override`) marks a method as
    overriding one declared on a base class. The base class's method is what
    calling code actually calls, so this override is used even though nothing
    calls it by its own name directly.
    """
    return match_many(decorator_names, _OVERRIDE_DECORATOR_NAMES)


def ignore_method(filename: Path, method_name: str) -> bool:
    return _is_special_name(method_name) or (
        (method_name in _PYTEST_METHOD_NAMES or method_name.startswith('test_')) and _is_test_file(filename)
    )


def is_self_attribute(node: ast.Attribute) -> bool:
    """Whether an attribute assignment target is `self.attr` (as opposed to
    e.g. `foo.attr`, where `foo` is some other object)."""
    return isinstance(node.value, ast.Name) and node.value.id == 'self'


def ignore_variable(filename: Path, varname: str) -> bool:
    """
    Ignore _ (Python idiom), _x (pylint convention) and
    __x__ (special variable or method), but not __x.
    """
    return (
        varname in IGNORED_VARIABLE_NAMES
        or (varname.startswith('_') and not varname.startswith('__'))
        or _is_special_name(varname)
    )

File: src/visitor/ignore.rs
//! Port of `deadcode/visitor/ignore.py`.

use std::path::Path;

use ruff_python_ast::Expr;

use crate::utils::fnmatch;
use crate::visitor::code_item::path_as_posix;

pub const IGNORED_VARIABLE_NAMES: &[&str] = &["object", "self"];
pub const PYTEST_FUNCTION_NAMES: &[&str] = &[
    "setup_module",
    "teardown_module",
    "setup_function",
    "teardown_function",
];
pub const PYTEST_METHOD_NAMES: &[&str] = &[
    "setup_class",
    "teardown_class",
    "setup_method",
    "teardown_method",
];
pub const PYTEST_FIXTURE_DECORATOR_NAMES: &[&str] =
    &["@pytest.fixture", "@pytest_asyncio.fixture", "@fixture"];
pub const PYTEST_USEFIXTURES_DECORATOR_NAMES: &[&str] = &[
    "@pytest.mark.usefixtures",
    "@mark.usefixtures",
    "@usefixtures",
];
pub const OVERRIDE_DECORATOR_NAMES: &[&str] = &[
    "@typing.override",
    "@typing_extensions.override",
    "@override",
];
/// Glob patterns, not literal names: the field name (`some_property` in
/// `@some_property.default`) is arbitrary user code, unrelated to how
/// `attrs`/`attr`'s `field`/`attrib`/`ib`/`define`/`attrs`/`s` were imported
/// or aliased — see `ignore_attrs_field_hook`.
pub const ATTRS_FIELD_HOOK_DECORATOR_PATTERNS: &[&str] = &["@*.default", "@*.validator"];

fn is_special_name(name: &str) -> bool {
    name.starts_with("__") && name.ends_with("__")
}

/// Port of `_is_test_file`. No `@lru_cache` here (unlike Python): our
/// `resolve_path` is a pure in-memory lexical normalization, not a
/// filesystem syscall, so there's no repeated-syscall cost to amortize.
pub fn is_test_file(filename: &Path) -> bool {
    let resolved = crate::utils::path_utils::resolve_path(filename);
    fnmatch::match_any(
        &path_as_posix(&resolved),
        &["*/test/*", "*/tests/*", "*/test*.py", "*[-_]test.py"],
        false,
    )
}

pub fn is_conftest_file(filename: &Path) -> bool {
    filename.file_name().and_then(|n| n.to_str()) == Some("conftest.py")
}

pub fn assigns_special_variable_all(targets: &[Expr]) -> bool {
    targets
        .iter()
        .any(|t| matches!(t, Expr::Name(n) if n.id.as_str() == "__all__"))
}

pub fn ignore_class(filename: &Path, class_name: &str) -> bool {
    is_test_file(filename) && class_name.contains("Test")
}

/// Ignore star-imported names (can't detect usage) and imports from
/// `__init__.py` files (commonly used to re-export/collect a package's API).
pub fn ignore_import(filename: &Path, import_name: &str) -> bool {
    filename.file_name().and_then(|n| n.to_str()) == Some("__init__.py") || import_name == "*"
}

pub fn ignore_function(filename: &Path, function_name: &str) -> bool {
    ((PYTEST_FUNCTION_NAMES.contains(&function_name) || function_name.starts_with("test_"))
        && is_test_file(filename))
        || ignore_pytest_hook(filename, function_name)
}

/// `pytest_*` functions in conftest.py (hooks pytest calls by name).
pub fn ignore_pytest_hook(filename: &Path, function_name: &str) -> bool {
    is_conftest_file(filename) && function_name.starts_with("pytest_")
}

/// Pytest fixtures are consumed by name-matching, never called directly, so
/// they'd otherwise look unused. Trust any fixture defined in conftest.py or
/// a recognized test file; fixtures elsewhere must be exempted explicitly.
pub fn ignore_pytest_fixture(filename: &Path, decorator_names: &[String]) -> bool {
    (is_conftest_file(filename) || is_test_file(filename))
        && fnmatch::match_many(decorator_names, PYTEST_FIXTURE_DECORATOR_NAMES, true)
}

/// `@typing.override` marks a method as overriding a base-class method — the
/// base's method is what's actually called, so this is used even though
/// nothing calls it by its own name.
pub fn ignore_override(decorator_names: &[String]) -> bool {
    fnmatch::match_many(decorator_names, OVERRIDE_DECORATOR_NAMES, true)
}

/// `@some_field.default` / `@some_field.validator` register an `attrs` field's
/// default-value factory or validator: `attrs` calls these itself, by the
/// field's own descriptor protocol, at class-creation/instantiation time —
/// never by the method's own name — so they'd otherwise look unused.
///
/// The field variable's name (`some_field`) never depends on how `field`/
/// `attrib`/`ib` (or the class decorator `define`/`attrs`/`s`) were imported
/// or aliased, so matching the decorator's shape alone — an attribute access
/// ending in `.default`/`.validator` — needs no import-alias tracking at all;
/// `some_field = attrs.field(...)`, `from attrs import field as fizz` +
/// `some_field = fizz(...)`, and every other spelling all produce the exact
/// same `@some_field.default`/`@some_field.validator` decorator.
pub fn ignore_attrs_field_hook(decorator_names: &[String]) -> bool {
    fnmatch::match_many(decorator_names, ATTRS_FIELD_HOOK_DECORATOR_PATTERNS, true)
}

pub fn ignore_method(filename: &Path, method_name: &str) -> bool {
    is_special_name(method_name)
        || ((PYTEST_METHOD_NAMES.contains(&method_name) || method_name.starts_with("test_"))
            && is_test_file(filename))
}

/// Whether an attribute assignment target is `self.attr` (as opposed to e.g.
/// `foo.attr`, where `foo` is some other object).
pub fn is_self_attribute_target(value: &Expr) -> bool {
    matches!(value, Expr::Name(n) if n.id.as_str() == "self")
}

/// Ignore `_` (Python idiom), `_x` (pylint convention), and `__x__` (special
/// variable/method), but not `__x`.
pub fn ignore_variable(varname: &str) -> bool {
    IGNORED_VARIABLE_NAMES.contains(&varname)
        || (varname.starts_with('_') && !varname.starts_with("__"))
        || is_special_name(varname)
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::path::PathBuf;

    #[test]
    fn test_file_detection() {
        assert!(is_test_file(&PathBuf::from("tests/test_foo.py")));
        assert!(is_test_file(&PathBuf::from("app/test_foo.py")));
        assert!(!is_test_file(&PathBuf::from("app/foo.py")));
    }

    #[test]
    fn conftest_detection() {
        assert!(is_conftest_file(&PathBuf::from(
            "tests/integration/conftest.py"
        )));
        assert!(!is_conftest_file(&PathBuf::from("tests/foo.py")));
    }

    #[test]
    fn variable_ignore_rules() {
        assert!(ignore_variable("_"));
        assert!(ignore_variable("_private"));
        assert!(ignore_variable("__dunder__"));
        assert!(!ignore_variable("__partial"));
        assert!(!ignore_variable("normal_name"));
    }

    #[test]
    fn attrs_field_hook_rules() {
        assert!(ignore_attrs_field_hook(&[
            "@some_property.default".to_string()
        ]));
        assert!(ignore_attrs_field_hook(&[
            "@some_property.validator".to_string()
        ]));
        // The field's own name is arbitrary — no fixed list to match against.
        assert!(ignore_attrs_field_hook(&["@fizz.default".to_string()]));
        assert!(!ignore_attrs_field_hook(&["@property".to_string()]));
        assert!(!ignore_attrs_field_hook(&[
            "@some_property.other".to_string()
        ]));
    }

    #[test]
    fn method_ignore_rules() {
        assert!(ignore_method(&PathBuf::from("foo.py"), "__init__"));
        assert!(ignore_method(
            &PathBuf::from("tests/test_foo.py"),
            "test_something"
        ));
        assert!(!ignore_method(
            &PathBuf::from("app/foo.py"),
            "test_something"
        ));
    }
}
